### PR TITLE
planner: Fixup error when assign `DEFAULT` in `INSERT/UPDATE/REPLACE ... SET ...` statement

### LINKED ISSUE #11939: Support using `DEAFULT` in `INSERT/UPDATE/REPLACE ... SET xxx=DEFAULT`  statement
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```
mysql> create table t (a int default 10);
mysql> update t set a=default;
```

2. What did you expect to see?
```
mysql> create table t (a int default 10);
Query OK, 0 rows affected (0.06 sec)

mysql> update t set a=default;
Query OK, 0 rows affected (0.01 sec)
Rows matched: 0  Changed: 0  Warnings: 0
```

3. What did you see instead?
```
mysql> create table t (a int default 10);
Query OK, 0 rows affected (0.00 sec)

mysql> update t set a=default;
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 22 near ""
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
TiDB v3.0.3
TiDB v2.1.16

### PR DESCRIPTION
**Pull request by Deardrops**
### What problem does this PR solve? 

Fix #11939 
```sql
create table t1 (a int default 1);
insert into t1 value (10);
update t1 set a = default; # <-- TiDB panic
replace t1 set a = default; # <-- TiDB panic
insert into t1 set a = default; # <-- TiDB panic
```

### What is changed and how it works?

In statement `update t1 set a = default`, When we build `Update` logic plan, we do:

1. For nongenerated columns, fill the `defaultExpr.Name` by `assign.Column` before rewrite its assignment expression.
2. For generated columns,
	1. If assigned `DEAFULT`, just ignore this column.
	2. If assigned any other value, throw an error `The value specified for generated column 'gc' in table 't' is not allowed.`

### Check List 

Tests 

 - Integration test

Code changes

 - None

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

Release note

 - Support using `DEAFULT` in `INSERT/UPDATE/REPLACE ... SET xxx=DEFAULT` statement

### CODE REVIEW

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -3005,7 +3005,7 @@ func (b *PlanBuilder) buildUpdateLists(
 	error error,
 ) {
 	b.curClause = fieldList
-	modifyColumns := make(map[string]struct{}, p.Schema().Len()) // Which columns are in set list.
+	modifyColumns := make(map[string]bool, p.Schema().Len()) // Which columns are in set list.
```
Comment (AilinKid) [inline planner/core/logical_plan_builder.go]: 1:better move the comment to the top of the line.
2:here seems can still use `map[string]struct{}`, `true` write it into map, `false` do nothing.

File: planner/core/logical_plan_builder.go [hunk 2]
```diff
@@ -3101,6 +3112,17 @@ func (b *PlanBuilder) buildUpdateLists(
 	return newList, p, allAssignmentsAreConstant, nil
 }
 
+// extractDefaultExpr extract a `DefaultExpr` without any parameter from a `ExprNode`
```
Comment (AilinKid) [inline planner/core/logical_plan_builder.go]: care about the punctuation and grammar.

File: planner/core/logical_plan_builder.go (hunk 1 above)
Comment (Deardrops) [reply planner/core/logical_plan_builder.go]: 1. OK, good idea.
2. In [L3044](https://github.com/pingcap/tidb/pull/13211/files#diff-b9e75a99170ea35a67074d5d3d3305adR3044), We need both indicate if this column is a modified column and if the value assigned to this column is `DEFAULT`. So, keep we use `map[string]bool`

**Comment by djshow832** [comment]
Did you test `insert ... on duplicate key update`?

**Comment by Deardrops** [comment]
> Did you test `insert ... on duplicate key update`?

@djshow832 Yes, In #13168, PTAL

File: planner/core/logical_plan_builder.go [hunk 3]
```diff
@@ -3074,6 +3087,17 @@ func (b *PlanBuilder) buildUpdateLists(
 	return newList, p, allAssignmentsAreConstant, nil
 }
 
+// extractDefaultExpr extract a `DefaultExpr` without any parameter from a `ExprNode`,
+// return it and if successful extract it.
+// Note: the SQL function `DEFAULT(a)` is not same with keyword `DEFAULT`,
```
Comment (djshow832) [inline planner/core/logical_plan_builder.go]:
```suggestion
// return the `DefaultExpr` and whether it's extracted successfully.
// Note: the SQL function `DEFAULT(a)` is not the same with keyword `DEFAULT`,
```

File: planner/core/logical_plan_builder.go (hunk 1 above)
Comment (AilinKid) [reply planner/core/logical_plan_builder.go]: Yeah, I mean, if you can find it in map, that means `DEFAULT`, if you can't find it / !ok, that means `NOT DEFAULT`.

File: executor/write_test.go [hunk 4]
```diff
@@ -1403,6 +1480,40 @@ func (s *testSuite8) TestUpdate(c *C) {
...
+	tk.MustQuery("select * from t2;").Check(testkit.Rows("10 -10 -10", "20 -20 -20"))
+	tk.MustExec("update t2 set a=30, b=default where a=10;")
+	tk.MustQuery("select * from t2;").Check(testkit.Rows("30 -30 -30", "20 -20 -20"))
+	tk.MustExec("update t2 set c=default, a=40 where c=-20;")
+	tk.MustQuery("select * from t2;").Check(testkit.Rows("30 -30 -30", "40 -40 -40"))
+	tk.MustExec("update t2 set a=default, b=default, c=default where b=-30;")
+	tk.MustQuery("select * from t2;").Check(testkit.Rows("1 -1 -1", "40 -40 -40"))
+	tk.MustExec("update t2 set a=default(a), b=default, c=default;")
+	tk.MustQuery("select * from t2;").Check(testkit.Rows("1 -1 -1", "1 -1 -1"))
+	tk.MustGetErrCode("update t2 set b=default(a);", mysql.ErrBadGeneratedColumn)
```
Comment (lamxTyler) [inline executor/write_test.go]: What about test for `update t2 set b=default(b);`? Seems that mysql allows it.

Comment (Deardrops) [reply executor/write_test.go]: @lamxTyler  We have discussed this case, and have agreement that only allow `default` keyword for generated column assignment. Use `default(a)` in here is disallowed. e.g.
```sql
create table t (a int, b int as (-a));
insert into t value (1, default);
update t set b = default;    # both allowed in MySQL and TiDB
update t set b = default(a); # allowed in MySQL, but disallowed in TiDB
update t set b = default(b); # allowed in MySQL, but disallowed in TiDB
```

Comment (lamxTyler) [reply executor/write_test.go]: Can you comment the reason that only allow `default` in code?

**Pull request merged**